packages/react/src/Form/Form.tsx

```diff
@@ -1,27 +1,41 @@
-import { CSSProperties, ReactNode, SyntheticEvent } from 'react';
+import { CSSProperties, ReactNode, SyntheticEvent, useState } from 'react';
+import { FormContext } from './Form.context';
 import { parseForm } from './FormUtils';
 
 export interface FormProps {
-  readonly onSubmit?: (formData: Record<string, string>) => void;
+  readonly onSubmit?: (formData: Record<string, string>) => Promise<void> | void;
   readonly style?: CSSProperties;
   readonly children?: ReactNode;
   readonly testid?: string;
 }
 
 export function Form(props: FormProps): JSX.Element {
+  const [submitting, setSubmitting] = useState(false);
   return (
-    <form
-      style={props.style}
-      data-testid={props.testid}
-      onSubmit={(e: SyntheticEvent) => {
-        e.preventDefault();
-        const formData = parseForm(e.target as HTMLFormElement);
-        if (props.onSubmit) {
-          props.onSubmit(formData);
-        }
-      }}
-    >
-      {props.children}
-    </form>
+    <FormContext.Provider value={{ submitting }}>
+      <form
+        style={props.style}
+        data-testid={props.testid}
+        onSubmit={(e: SyntheticEvent) => {
+          e.preventDefault();
+          const formData = parseForm(e.target as HTMLFormElement);
+          if (props.onSubmit) {
+            setSubmitting(true);
+            const result = props.onSubmit(formData);
+            if (result?.then) {
+              result
+                .then(() => {
+                  setSubmitting(false);
+                })
+                .catch(console.error);
```

Comment:
If there's an error, should the button be returned to the normal state so the user can attempt to re-submit?

Reply:
Yes, good catch, will change to 'finally'